apps/ex_wire/lib/ex_wire/sync.ex

```diff
@@ -161,39 +181,43 @@ defmodule ExWire.Sync do
           block_queue: block_queue,
           block_tree: block_tree,
           chain: chain,
-          trie: trie
+          trie: trie,
+          highest_block_number: highest_block_number
         }
       ) do
-    {next_block_queue, next_block_tree, next_trie, header_hashes} =
-      Enum.reduce(block_headers.headers, {block_queue, block_tree, trie, []}, fn header,
-                                                                                 {block_queue,
-                                                                                  block_tree,
-                                                                                  trie,
-                                                                                  header_hashes} ->
-        header_hash = header |> Header.hash()
-
-        {next_block_queue, next_block_tree, next_trie, should_request_block} =
-          BlockQueue.add_header(
-            block_queue,
-            block_tree,
-            header,
-            header_hash,
-            peer.remote_id,
-            chain,
-            trie
-          )
-
-        next_header_hashes =
-          if should_request_block do
-            :ok = Logger.debug(fn -> "[Sync] Requesting block body #{header.number}" end)
-
-            [header_hash | header_hashes]
-          else
-            header_hashes
-          end
-
-        {next_block_queue, next_block_tree, next_trie, next_header_hashes}
-      end)
+    {next_highest_block_number, next_block_queue, next_block_tree, next_trie, header_hashes} =
+      Enum.reduce(
+        block_headers.headers,
+        {highest_block_number, block_queue, block_tree, trie, []},
+        fn header, {highest_block_number, block_queue, block_tree, trie, header_hashes} ->
+          header_hash = header |> Header.hash()
+
+          {next_block_queue, next_block_tree, next_trie, should_request_block} =
+            BlockQueue.add_header(
+              block_queue,
+              block_tree,
+              header,
+              header_hash,
+              peer.remote_id,
+              chain,
+              trie
+            )
+
+          next_header_hashes =
+            if should_request_block do
+              :ok = Logger.debug(fn -> "[Sync] Requesting block body #{header.number}" end)
+
+              [header_hash | header_hashes]
+            else
+              header_hashes
+            end
+
+          next_highest_block_number = Integer.max(highest_block_number, header.number)
```

Comment:
@ayrat555 would love your opinion on this. is this the correct way of estimating the highest block number?

Reply:
Looks like I lead you astray. It's `Kernel.max(...)`